### ISSUE TITLE
Handlers not ran when first skipped but then notified

### ISSUE THREAD

**Issue by arturhoo** [ansible/ansible#11327]
**Issue Type:**

Bug Report

**Ansible Version:**

```
ansible 1.9.2 (detached HEAD 982bad7886) last updated 2015/06/19 17:29:23 (GMT -200)
  lib/ansible/modules/core: (detached HEAD 016ad8a1cb) last updated 2015/06/19 17:29:29 (GMT -200)
  lib/ansible/modules/extras: (detached HEAD 86d9bbae52) last updated 2015/06/19 17:29:34 (GMT -200)
  v2/ansible/modules/core: (detached HEAD 34784b7a61) last updated 2015/06/19 17:29:40 (GMT -200)
  v2/ansible/modules/extras: (detached HEAD 46e316a20a) last updated 2015/06/19 17:29:47 (GMT -200)
  configured module search path = None
```

**Ansible Configuration:**

I haven't changed or added any configuration settings to Ansible.

**Environment:**

OSX 10.10.3
Python 2.7.10

**Summary:**

A handler that is firstly not notified due to a skipped and depended upon task, but later notified through the same task (directly invoked), does not run at the of the play.

**Steps to Reproduce:**

I set up an example repository to illustrate this bug: https://github.com/arturhoo/ansible-bug

**Expected Results:**

```
$ ansible-playbook -i 'localhost,' --connection=local playbook.yml

PLAY [all] ********************************************************************

GATHERING FACTS ***************************************************************
ok: [localhost]

TASK: [one | create bar file] *************************************************
skipping: [localhost]

TASK: [one | echo foo to bar file] ********************************************
skipping: [localhost]

TASK: [two | create xpto file] ************************************************
changed: [localhost]

TASK: [one | create bar file] *************************************************
changed: [localhost]

TASK: [one | echo foo to bar file] ********************************************
changed: [localhost]

NOTIFIED: [one | handle this] *************************************************
changed: [localhost]

PLAY RECAP ********************************************************************
localhost                  : ok=4    changed=4    unreachable=0    failed=0
```

**Actual Results:**

```
$ ansible-playbook -i 'localhost,' --connection=local playbook.yml

PLAY [all] ********************************************************************

GATHERING FACTS ***************************************************************
ok: [localhost]

TASK: [one | create bar file] *************************************************
skipping: [localhost]

TASK: [one | echo foo to bar file] ********************************************
skipping: [localhost]

TASK: [two | create xpto file] ************************************************
changed: [localhost]

TASK: [one | create bar file] *************************************************
changed: [localhost]

TASK: [one | echo foo to bar file] ********************************************
changed: [localhost]

NOTIFIED: [one | handle this] *************************************************
skipping: [localhost]

PLAY RECAP ********************************************************************
localhost                  : ok=4    changed=3    unreachable=0    failed=0
```

**Comment by bcoca** [comment]
I'm guessing this has to do with multiple imports of the same role?

**Comment by arturhoo** [comment]
Well, I don't know if I understood you correctly, but role `two` depends on role one based on a variable:

> `roles/two/meta/main.yml`

``` yml
---
dependencies:
  - { role: one, when: dependant }
```

As you can see on the playbook, this variable is set to false and role `one` is firstly skipped, only to be executed later.

**Comment by jimi-c** [comment]
Just a note, this is still an issue in devel (v2), however there's no good way to solve this currently. The main reason comes down to the fact that with role de-duplication, we either create an internal copy of the role when the tags/conditionals are different or we don't. If we do, then this causes an issue with handlers because there are now multiple handlers with the same name. If we don't, then the conditionals are evaluated incorrectly (due to the role having multiple parents, aka a "diamond" pattern).

The example given above is really a bad example, as you should never have a role at multiple levels like this (both at the top level and as a dep). This will however affect roles using handlers where `allow_duplicates: true` is used, so we'll hold this open for the time being.

**Comment by davidbirdsong** [comment]
> The example given above is really a bad example, as you should never have a role at multiple levels like this (both at the top level and as a dep). This will however affect roles using handlers where allow_duplicates: true is used, so we'll hold this open for the time being.

I'm curious why one should never have roles sprinkled around wherever needed. Aren't roles supposed to be the nugget in ansible that promotes re-use?

It took me a long time to pare down my use of roles and discover simple, easy-to-reason-about roles that can be re-used over and over again by supplying vars that overwrite role defaults. To say that handlers can't fire for each invocation of that role really undermines re-use and roles fall flat.

**Comment by jimi-c** [comment]
FWIW, this is actually (mostly) fixed in the current devel/stable-2.1 branches (and most likely 2.0). The only catch right now is that the handler looks for parent variables in roles and fails to find `dependent` when notified from role `one` in the example. This is because it is improperly looking at all parent roles rather than the dependency chain it came from.

However, this is a different bug (opened as https://github.com/ansible/ansible/issues/15884), and as such we'll close this one now. If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
